Update Code Review/Pull Request Guidelines (#803)
department-of-veterans-affairs/va.gov-team: Practice Areas/Engineering/code_review_guidelines.md

## Guidelines

* **Review lifecycle**
    * Before sending code to VSP team have it reviewed by your team (see **First review** below) using a [Draft Pull Request](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests).  
      * This prevents [code owners](https://help.github.com/en/articles/about-code-owners) from being notified before pull request is ready for review.

>     * After having your pull request reviewed by your team you can mark your pull request as ready for review.

Reviewer: Might be worth calling out how to do this. Is there a button to graduate a draft PR to a full-fledged PR?

Author: There is, 
<img width="680" alt="Screen Shot 2019-08-07 at 2 19 41 PM" src="https://user-images.githubusercontent.com/1094999/62647383-6b305d80-b91e-11e9-964b-7ca41a85bc5b.png">

Author: updating section to indicate this

Author: also added link to https://github.blog/2019-02-14-introducing-draft-pull-requests/

      * An engineer from the VSP will aim to review within 24 hours (business days) and each additional round of changes will be reviewed within 24 hours.  If you need it faster, escalate via your DSVA product owner.
      * Please **do not post in Slack** channels requesting code reviews unless it has been more then 24 hours (business days).
    * After sending your code for review, don't add additional changes! Code needs to be stable for a safe review.

>     * If non-trivial changes to your code are made *after approval by VSP team*, please notify approvers so they can re-review your pull request.

Reviewer: I like this better, but I'm not sure if we should nix the bit about reverting the change if there was code that was added / changed after the approval.

@gunsch What do you think?

Reviewer 2: Please leave the original wording. The point is to clearly indicate what action VSP team will take if non-trivial changes are submitted after review.

If you want to see that clarified better, I'd prefer:

```suggestion
    * If non-trivial changes to your code are made *after approval by VSP team* and submitted without a re-review, your change is subject to being reverted.
```

Author: made suggested change

## Guidelines

* **Review lifecycle**
    * Before sending code to VSP team have it reviewed by your team (see **First review** below) using a [Draft Pull Request](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests).  
      * This prevents [code owners](https://help.github.com/en/articles/about-code-owners) from being notified before pull request is ready for review.

>       * [Learn more about draft pull requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

Reviewer 2: let's drop this line, too -- line 14 already links to the other Draft PR doc, and line 25 has this link too

Author: removed and changed link above to this link, as https://github.blog/2019-02-14-introducing-draft-pull-requests/ is more helpful with how to use draft pull requests

    * After having your pull request reviewed by your team you can click the *Ready for review* button in the merge section of your pull request.

>       * An engineer from the VSP will aim to review within 24 hours (business days) and each additional round of changes will be reviewed within 24 hours.  If you need it faster, escalate via your DSVA product owner.

Reviewer 2: ```suggestion
      * An engineer from the VSP team will aim to review within 24 hours (business days) and each additional round of changes will be reviewed within 24 hours.  If you need a review faster, escalate via your DSVA product owner.
```

>       * Please **do not post in Slack channels requesting VSP code reviews** unless it has been more then 24 hours (business days).

Reviewer 2: ```suggestion
      * **Do not post in Slack channels requesting VSP code reviews** unless it has been more than 24 hours since requesting review on GitHub.
```

    * After sending your code for review, don't add additional changes! Code needs to be stable for a safe review.
    * If non-trivial changes to your code are made *after approval by VSP team* and submitted without a re-review, your change is subject to being reverted.

* **First review**: 
  * Use a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

>   * For *all code* tag your PR with your team-specific label(s)

Reviewer 2: ```suggestion
  * Tag your PR with your team-specific label(s)
```

>   * Your code should first be reviewed and approved by a member of your team. 

Reviewer 2: ```suggestion
  * For this first review, your code should be reviewed and approved by a member of your team. 
```

## Guidelines

* **Review lifecycle**
    * Before sending code to VSP team have it reviewed by your team (see **First review** below) using a [Draft Pull Request](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests).  

>       * This prevents [code owners](https://help.github.com/en/articles/about-code-owners) from being notified before pull request is ready for review.

Reviewer: I dig it! This is awesome.

    * After having your pull request reviewed by your team you can mark your pull request as ready for review.

>       * An engineer from the VSP will aim to review within 24 hours (business days) and each additional round of changes will be reviewed within 24 hours.  If you need it faster, escalate to your lead or scrum master.

Reviewer: I like the idea of going through the lead or scrum master, but I think it still ultimately needs to go through Leah or Patrick.

Does that sound right, @gunsch?

Reviewer 2: to leave the wording more general, let's say "escalate via your DSVA product owner"

Author: changed

>       * Please **do not post in Slack** channels requesting code reviews unless it has been more then 24 hours (business days).

Reviewer: :100:

    * After sending your code for review, don't add additional changes! Code needs to be stable for a safe review.
    * If non-trivial changes to your code are made *after approval by VSP team*, please notify approvers so they can re-review your pull request.

* **First review**: 
  * Use a [Draft Pull Request](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests)

>   * Add your team specific label(s)

Reviewer 2: ```suggestion
  * Tag your PR with the following labels:
    * For *all code*, add your team-specific label(s)
    * For *database migrations*, add the `database-migration` label
    * For *dependency/package upgrades*, add the `package-change` label
```

>     * Additionally if your change requires:

Reviewer: Probably doesn't need to be a sub-point of the team labels item.

>       * A database migration please use the `database-migration` label

Reviewer: Is this how the `database-migration` label is being used right now?

Reviewer 2: cc @annaswims

Author: The BAH team has started using it in https://github.com/department-of-veterans-affairs/gibct-data-service to allow reviewers to more easily know if a PR contains database changes

>       * A dependency/package upgrade please use the `package-change` label

Reviewer: Does this label exist right now?

Reviewer 3: This label doesn't exist in vets-api.  Can you explain the need for this label?

Author: Easier for reviewers to know at a glance when PRs contain database or package/dependency changes

Author: Created the labels for 
- https://github.com/department-of-veterans-affairs/gibct-data-service
- https://github.com/department-of-veterans-affairs/vets-api
- https://github.com/department-of-veterans-affairs/vets-website

Reviewer 2: Hmm, I don't think we need labels for that though. If a PR contains database or package changes, call that out in the issue heading and/or PR description?

I'll leave this up to @annaswims to make a final call on whether or not we want this, but in general we have a preference _against_ adding extra process until there's a clear need for it.

Reviewer: I agree with @gunsch. I think this kind of indication can be added in the PR description as needed.

In general, PR descriptions should be pretty thorough, so it seems like a good fit there.

Author: Which would be better?
```
  * Tag your PR with the following labels:
    * For *all code*, add your team-specific label(s)
    * (Optional) For *database migrations*, add the `database-migration` label
    * (Optional) For *dependency/package upgrades*, add the `package-change` label
```
or
```
  * Tag your PR with the following labels:
    * For *all code*, add your team-specific label(s)
```

    * **PR title** should be a concise summary of the change.
    * In your **PR description**, discuss how code changes are related to the pull request objectives.
    * Reference the **GitHub issues** your PR relates to, in order to provide reviewers more context.

>     * If the pull request is a dependency/package upgrade or database migration for another pull request please reference that pull request.

Reviewer: This _sounds_ right, but I'm not familiar with the process myself.

@annaswims What're your thoughts?

Reviewer 3: I think this is generally covered in `discuss how code changes are related to the pull request objectives`

Author: removed

* **Size/complexity**: keep your changes scoped to one feature at a time. This means:

>     * **Size**: changes should be smaller than **500** changed lines (not including tests or whitespace). *If your change can't be smaller than this, talk to the team in #vfs-platform-support before asking for a review.*

Reviewer: Let's stick to 300 for now.
